compiler/src/dotty/tools/dotc/ast/tpd.scala

```diff
@@ -730,6 +730,27 @@ object tpd extends Trees.Instance[Type] with TypedTreeInfo {
       if (from == to) tree else loop(from, Nil, to :: Nil)
     }
 
+    /**
+     * Set the owner of every definition in this tree which is not itself contained in this
+     * tree to be `newowner`
+     */
+    def changeNonLocalOwners(newowner: Symbol)(implicit ctx: Context): Tree = {
+      val localOwners = mutable.HashSet[Symbol]()
+      val traverser = new TreeTraverser {
+
+        def traverse(tree: Tree)(implicit ctx: Context): Unit = {
+          tree match {
+            case _: DefTree => if(tree.symbol ne NoSymbol) localOwners += tree.symbol
+            case _ =>          traverseChildren(tree)
+          }
+        }
+      }
+      traverser.traverse(tree)
+      val nonLocalOwners = localOwners.filter(sym => !localOwners.contains(sym.owner)).toList.map(_.owner)
+      val newOwners = nonLocalOwners.map(_ => newowner)
+      new TreeTypeMap(oldOwners = nonLocalOwners, newOwners = newOwners).apply(tree)
+    }
```

Comment:
Can this be simplified further?
```scala
def changeNonLocalOwners(newOwner: Symbol)(implicit ctx: Context): Tree = {
  val localOwnerAcc = new TreeAccumulator[List[Symbol]] {
    def apply(ss: List[Symbol], tree: Tree)(implicit ctx: Context) = tree match {
      case tree: DefTree =>
        if (tree.symbol.exists) tree.symbol :: ss
        else ss
      case _ =>
        foldOver(ss, tree)
      }
    }
  }
  val nonLocalOwners = localOwnerAcc(Nil, tree).map(_.owner).distinct
  val newOwners = nonLocalOwners.map(_ => newOwner)
  new TreeTypeMap(oldOwners = nonLocalOwners, newOwners = newOwners).apply(tree)
}
```

Reply:
I think the idea of using an accumulator is good, but why the detour via lists? Let's use an accumulator over an immutable set instead.

Reply:
Updated:
```scala
def changeNonLocalOwners(newOwner: Symbol)(implicit ctx: Context): Tree = {
  val ownerAcc = new TreeAccumulator[List[Symbol]] {
    def apply(ss: Set[Symbol], tree: Tree)(implicit ctx: Context) = tree match {
      case tree: DefTree =>
        if (tree.symbol.exists) tree.symbol + ss.owner
        else ss
      case _ =>
        foldOver(ss, tree)
      }
    }
  }
  val owners = ownerAcc(Nil, tree)
  val newOwners = List.fill(owners.size)(newOwner)
  new TreeTypeMap(oldOwners = owners, newOwners = newOwners).apply(tree)
}
```